### PR TITLE
Handle exceptions in underlying middleware when tracking response size

### PR DESCRIPTION
**Pull request by jason-o-matic**

### CODE REVIEW

File: lib/metrician/middleware/request_timing.rb [hunk 1]
```diff
@@ -143,7 +143,7 @@ def self.extract_route(controller:, path:)
       end
 
       def self.get_response_size(headers:, body:)
-        return headers[HEADER_CONTENT_LENGTH] if headers[HEADER_CONTENT_LENGTH]
+        return headers[HEADER_CONTENT_LENGTH] if headers && headers[HEADER_CONTENT_LENGTH]
```
Comment (janxious) [inline lib/metrician/middleware/request_timing.rb:146]: Is this related to the error?

Comment (jason-o-matic) [reply lib/metrician/middleware/request_timing.rb:146]: Yeah, the cause of the error was headers being nil.

Comment (janxious) [reply lib/metrician/middleware/request_timing.rb:146]: Is there a way to test that without raising an exception in the test? Or is the only time it happens?

Comment (jason-o-matic) [reply lib/metrician/middleware/request_timing.rb:146]: I think that's the only time it happens. An exception on http://github.com/Instrumental/metrician/blob/2d7c8b68b0ede4412d873ae15e0272088ed071ee/lib/metrician/middleware/request_timing.rb#L33 will cause the variables to not be set.

File: spec/metrician_spec.rb [hunk 2]
```diff
@@ -271,6 +271,28 @@ def app
...
+        require "metrician/middleware/application_timing"
+        Rack::Builder.app do
+          use Metrician::Middleware::RequestTiming
+          use Metrician::Middleware::ApplicationTiming
+          run lambda { |env| raise "boom" }
+        end
+      end
+
+      specify "middleware exceptions don't cause errors in response size tracking" do
+        Metrician.configuration[:request_timing][:response_size][:enabled] = true
```
Comment (janxious) [inline spec/metrician_spec.rb:286]: I been thinking about how this affects downstream tests, and I think we should have a thing in our test setup that resets the configuration to default after every test. I don't have an expectation that will happen on this PR.

File: lib/metrician/middleware/request_timing.rb [hunk 3]
```diff
@@ -51,7 +51,7 @@ def call(env)
             if Middleware.error?
               # We to_i the status because in some circumstances it is a
               # Fixnum and some it is a string. Because why not.
-              increment(:error, current_route) if status.to_i >= 500
+              increment(:error, current_route) if $! || status.to_i >= 500
```
Comment (janxious) [inline lib/metrician/middleware/request_timing.rb]: We've had `$ERROR_INFO` since like ruby 1998. You should use that, as it is a more scrutable thing.

https://github.com/ruby/ruby/blob/trunk/lib/English.rb#L51

**Review by janxious — CHANGES_REQUESTED** [review]

File: spec/metrician_spec.rb (hunk 2 above)
Comment (jason-o-matic) [reply spec/metrician_spec.rb:286]: `Metrician.instance_variable_set("@configuration", nil)`?

File: lib/metrician/middleware/request_timing.rb (hunk 3 above)
Comment (jason-o-matic) [reply lib/metrician/middleware/request_timing.rb]: Done.

File: spec/metrician_spec.rb (hunk 2 above)
Comment (janxious) [reply spec/metrician_spec.rb:286]: Yeah, that would probably do it

**Review by janxious — APPROVED** [review]

File: spec/metrician_spec.rb (hunk 2 above)
Comment (jason-o-matic) [reply spec/metrician_spec.rb:286]: Eh, it's complicated because of all the instance variable memoization. I don't have a great next idea so I'll punt for now.

Comment (janxious) [reply spec/metrician_spec.rb:286]: I decided to do this and took this approach: https://github.com/Instrumental/metrician-ruby/commit/6a69af92af287af5ef377c0dfb29b601bbd6c66a 

Feel free to yank that if you think it is a bad one.
